### PR TITLE
bluetooth: fix gatt_dm discovery of empty services

### PR DESCRIPTION
**Pull request by dchat-nordic**
Empty services (no characterictics) can be discovered.

Jira:  NCSDK-13296

Signed-off-by: Dominik Chat <dominik.chat@nordicsemi.no>

### CODE REVIEW

**Review by peknis — APPROVED** [review]

File: subsys/bluetooth/gatt_dm.c [hunk 1]
```diff
@@ -290,9 +290,13 @@ static uint8_t discovery_process_service(struct bt_gatt_dm *dm,
 		return BT_GATT_ITER_STOP;
 	}
 
-	LOG_DBG("Service detected, handles range: <%u, %u>",
-		cur_attr->handle + 1,
-		service_val->end_handle);
+	if (cur_attr->handle == service_val->end_handle) {
+		LOG_DBG("Empty service detected with handle: %u", cur_attr->handle);
```
Comment (KAGA164) [inline subsys/bluetooth/gatt_dm.c:294]: What happens when we don't return here. It looks like there is a risk that we will start  attribute discovery with invalid handle range. Am I right or I missed something ?

**Review by KAGA164 — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/gatt_dm.c [hunk 2]
```diff
@@ -311,6 +315,7 @@ static uint8_t discovery_process_service(struct bt_gatt_dm *dm,
 
 	if (cur_attr->handle == cur_service_val->end_handle) {
 		/* No characteristics to discover, go to next service. */
+		discovery_complete(dm);
```
Comment (KAGA164) [inline subsys/bluetooth/gatt_dm.c:318]: Wouldn't be better to call this callback in next iteration when no other service with the same UUID is found ?

Comment (dchat-nordic) [reply subsys/bluetooth/gatt_dm.c:318]: Can you provide the scenario when this method could fail?

Comment (KAGA164) [reply subsys/bluetooth/gatt_dm.c:318]: There can be corner case situation when there are two the same service instances and the first one is empty, then would be good to discover a second one, but this scenario is very unlikely.

**Review by KAGA164 — APPROVED** [review]

**Review by grochu — COMMENTED** [review]
The code looks good, but perhaps expanding the module test case is required to ensure coverage of the new scenario: https://github.com/nrfconnect/sdk-nrf/tree/main/tests/subsys/bluetooth/gatt_dm?

**Comment by gopiotr** [comment]
From test point of view it looks okay. I tested this handling of empty services also on BabbleSim and it works properly. Thank you for adding tests to https://github.com/nrfconnect/sdk-nrf/tree/main/tests/subsys/bluetooth/gatt_dm :)

**Review by gopiotr — APPROVED** [review]

**Pull request merged**